### PR TITLE
EMA Self-Distillation: use EMA predictions as soft targets

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The baseline uses EMA (decay=0.999) to maintain a smoothed copy of the model for evaluation. The EMA model's predictions are typically more stable and better-generalized than the online model's, because exponential averaging smooths out training noise.

**Self-distillation** adds a secondary loss that pushes the online model's predictions toward the EMA model's predictions. This creates a feedback loop:
- EMA model: slow-moving, stable, provides "consensus" predictions
- Online model: learns from both ground truth AND EMA's smoothed predictions
- The distillation loss acts as a **confidence regularizer** — it penalizes the online model when its predictions diverge significantly from the EMA's stable predictions

**Why it might work here:**
1. **Implicit ensemble effect:** The EMA model averages over many training steps. Distilling from it is like distilling from a temporal ensemble.
2. **OOD regularization:** On OOD samples, the online model may produce noisy predictions that the EMA has smoothed over many steps. The distillation loss keeps the online model from making wild OOD predictions.
3. **Gradient smoothing:** The MSE distillation loss provides a SMOOTH gradient signal (unlike L1 which has discontinuous gradients). This complements the L1 primary loss.
4. **Zero extra compute:** The EMA model is already maintained in the baseline. We just need one extra forward pass (which is fast since it's in `torch.no_grad()`).

**Related work:** Born Again Networks (Furlanello et al., ICML 2018), Mean Teacher (Tarvainen & Valpola, NeurIPS 2017). In those works, using a teacher that tracks the student significantly improves generalization.

**Confidence:** Medium. Novel application to CFD surrogate training. The circular dependency (EMA tracks online, online distills from EMA) is stable at low weights. Main risk: if EMA and online model agree closely (which they should by late training), the distillation loss provides negligible signal.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
ema_self_distill: bool = False      # distill online model from EMA predictions
distill_weight: float = 0.1         # weight for distillation loss
distill_start_epoch: int = 20       # don't distill in very early training (EMA is not yet stable)
```

### Step 2: Add distillation loss in training loop

In the training step, AFTER the online model's forward pass and BEFORE the backward pass:

```python
if cfg.ema_self_distill and epoch >= cfg.distill_start_epoch:
    with torch.no_grad():
        # Forward pass through EMA model (no gradients needed)
        ema_pred = ema_model(x_input)  # same input as online model
        # ema_pred should have the same structure as online model output
        # (pressure predictions, velocity predictions, etc.)
    
    # MSE distillation loss on surface pressure predictions
    distill_loss = F.mse_loss(pred_p[surf_mask], ema_pred_p[surf_mask])
    loss = loss + cfg.distill_weight * distill_loss
    
    # Log to W&B
    if wandb_log:
        log_dict['train/distill_loss'] = distill_loss.item()
```

**IMPORTANT details:**
- Use `torch.no_grad()` for the EMA forward pass — we don't want gradients flowing through the EMA model
- Only distill the SURFACE PRESSURE predictions (not volume, not velocity) — surface pressure is what we optimize for
- Start distillation after epoch 20 — the EMA model needs time to accumulate meaningful averages before being used as a teacher
- Use MSE (not L1) for the distillation loss — MSE provides smoother gradients for matching predictions

### Step 3: Handle EMA forward pass

The EMA model should already be available (it's maintained for evaluation). You'll need to run a forward pass through it during training. This is straightforward since the EMA model has the same architecture as the online model.

If the EMA model is wrapped differently (e.g., not compiled, different module structure), ensure the forward pass produces compatible output tensors.

### Step 4: torch.compile compatibility

The EMA forward pass is in `torch.no_grad()` context and doesn't participate in the backward graph. It should be compile-safe. If the compiled model has issues with the EMA call, run the EMA forward outside the compiled region.

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent thorfinn --wandb_name "thorfinn/ema-distill-s42" \
  --wandb_group "round19/ema-self-distillation" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --ema_self_distill --distill_weight 0.1 --distill_start_epoch 20

# Seed 73 — identical but --seed 73 --wandb_name "thorfinn/ema-distill-s73"
```

### Step 6: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

Also report:
- `train/distill_loss` curve — does the online model converge toward the EMA? (loss should decrease over training)
- Comparison of EMA model vs online model metrics at final epoch
- Any impact on training speed (the extra EMA forward pass should be minimal)

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### EMA Self-Distillation — Neutral/Slightly Negative Result

**Summary:** EMA self-distillation produces results within noise of the baseline. The 2-seed average narrowly misses beating the baseline on all metrics, though individual seeds beat on specific splits.

### Metrics (Surface Pressure MAE, Pa)

| Metric | Baseline (2-seed avg) | Distill s42 | Distill s73 | Distill 2-seed avg | Delta |
|--------|----------------------|-------------|-------------|---------------------|-------|
| p_in   | **11.979** | 12.1 | 12.0 | 12.050 | **+0.6%** |
| p_tan  | **28.341** | **28.0** | 28.8 | 28.400 | **+0.2%** |
| p_oodc | **7.643**  | 7.9  | 7.8  | 7.850  | **+2.7%** |
| p_re   | **6.300**  | **6.2**  | 6.7  | 6.450  | **+2.4%** |

### Val Loss

| Split | S42 | S73 |
|-------|-----|-----|
| val_in_dist | 0.2433 | 0.2534 |
| val_tandem_transfer | 0.8673 | 0.8663 |
| val_ood_cond | 0.1964 | 0.1988 |
| val_ood_re | 0.1890 | 0.2002 |

### Experiment Details

- **Seed 42:** W&B run `0o8cofv5`, 146 epochs, 181.5 min, 46.6GB VRAM
- **Seed 73:** W&B run `kvs1oevx`, 146 epochs, 181.7 min, 46.4GB VRAM
- **Config:** `--ema_self_distill --distill_weight 0.1 --distill_start_epoch 20`
- EMA model (decay=0.999) runs a `torch.no_grad()` forward pass each training step after epoch 20
- MSE distillation loss on surface pressure predictions only (channel 2)
- Distillation loss added as shared term to PCGrad 3-way groups
- Extra EMA forward pass adds ~40s/epoch overhead (72→116s for later epochs when distillation active)

### Command Used

```bash
cd cfd_tandemfoil && python train.py \
  --agent thorfinn --wandb_name "thorfinn/ema-distill-s42" \
  --wandb_group "round19/ema-self-distillation" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --ema_self_distill --distill_weight 0.1 --distill_start_epoch 20
```

### What Happened

The distillation loss provides a valid gradient signal — `train/distill_loss` decreases over training, showing the online model converges toward the EMA's predictions. However, this convergence provides **no net benefit** over the baseline.

**Why it was neutral:**

1. **EMA and online model agree closely by mid-training.** With decay=0.999, the EMA is already very close to the online model. The distillation loss becomes near-zero quickly, providing negligible gradient signal when it matters most (late training, where fine-tuning happens).

2. **Redundancy with existing EMA evaluation.** The baseline already uses the EMA model for evaluation. Self-distillation tries to make the online model match the EMA, but the online model's gradient updates are what drive the EMA's parameters — there's a circular dependency that dampens the regularization effect.

3. **Compute cost not justified.** The extra EMA forward pass adds ~60% overhead per epoch after distillation starts (72s→116s). For a neutral result, this is wasted compute. Note: both seeds still managed 146 epochs within the 180-minute timeout, but with a tighter budget this overhead would reduce the number of training epochs.

4. **Surface-only distillation may be too narrow.** Distilling only on surface pressure (per the instructions) means the loss doesn't regularize volume predictions or velocity channels. A full-field MSE distillation might have more effect, but at the cost of even more gradient noise.

### Suggested Follow-ups

1. **Not recommended at distill_weight=0.1.** The effect is too small. Higher weights (0.5+) risk destabilizing training via the circular dependency.
2. **Delayed higher decay:** Using a very high decay (0.9999) with delayed start (epoch 40+) might create a more distinct teacher, but the circular dependency remains.
3. **External teacher distillation** (training a separate teacher model first, then distilling from it) avoids the circular dependency but doubles compute cost. Not practical in our setup.
4. **Focus elsewhere:** The baseline is already well-regularized with EMA eval, augmentations, and PCGrad. Additional regularization approaches face diminishing returns. Architecture/conditioning changes are more likely to yield gains.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2239 because ADVISOR: Closing — neutral result, all metrics marginally worse (p_in +0.6%, p_oodc +2.7%, p_tan +0.2%, p_re +2.4%). Seed 42 beat baseline on p_tan and p_re individually, but the 2-seed average misses. Your analysis about the circular dependency between online model and EMA teacher is insightful — the EMA is already so close to the online model that the distillation signal is near-zero when it matters most (late training). The 60% compute overhead for a neutral result confirms this direction isn't productive. Your suggestion that architecture/conditioning changes are more likely to yield gains aligns with the broader trend: the baseline training dynamics are near-optimal..
